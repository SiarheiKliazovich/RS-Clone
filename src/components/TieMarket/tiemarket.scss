@import "../../sass/variables";
@import "../../sass/mixins";

.overlay { 
  display: block;
  position: fixed;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  background: $color-black4;
  z-index: 19;
}

.tiemarket-wrapper {
  background: $color-grey-tie;
}

.tiemarket__banner-wrapper {
  background: $color-grey-tie;
  padding-top: 40px;
  padding-bottom: 40px;
}

.market-block {
  height: calc(100% - 30px);
  border: none;
  background:  $color-white1;
  border-radius: 5px;
}

.market-block__title_wrapper {
  background: transparent;
  border-radius: 0;
  padding: 20px;
  margin-bottom: 0;
  border-bottom: 2px solid $color-grey-tie;
  text-align: center;
  display: flex;
  justify-content: center;
}

.market-block__title {
  font-family: $font-family-base;
  font-weight: $font-weight-bold;
  line-height: 1;
  font-size: $font-size-llg;
  color: $color-black1;
}

.market-block__products {
  padding: 30px 20px;
}

.row__products {
  display: flex;
  flex-wrap: wrap;
  gap: 10px;

  @include media-small-mobile {
    flex-direction: column;
    align-items: center;
  }
}

.tie__product {
  display: flex;
  width: calc(25% - 10px);
  background: #f8f6f691;
  align-items: flex-start;

    @include media-tablet {
      width: calc(33% - 10px);
    }

    @include media-small-tablet {
      width: calc(50% - 10px);
    }

    @include media-mobile {
      width: 100%;
    }
}

.tie__products_img {
  cursor: pointer;
  display: flex;
}

.products_img {
  width: 115px;
  height: 190px;
}

.tie__products_discription {
  display: flex;
  flex-direction: column;
  padding-left: 10px;
  width: 100%;
  overflow: hidden;
  text-align: left;
}

.tie__products_name {
  padding-bottom: 10px;
  padding-top: 15px;
  font-family: $font-family-base;
  font-weight: $font-weight-bold;
  line-height: 1;
  font-size: $font-size-lg;
  color: $font-color-header-dark;
}

.tie__products_price {
  padding-bottom: 10px;
  font-family: $font-family-base;
  font-weight: $font-weight-bold;
  line-height: 1;
  font-size: $font-size-lg;
  color: $font-color-header-dark;
}

.tie__products_wrapper {
  padding-top: 22px;
  padding-bottom: 10px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.tie__products_btn {
  border: none;
  border-radius: 3px;
  color: $color-white1;
  transition: 0.4s ease-in-out;
  padding: 10px;
  cursor: pointer;
  min-width: 125px;
  text-align: left;

  .fa {
    margin-right: 10px;
    color: $color-white1;
  }
}

.btn__like {
  background-repeat: no-repeat;
  border: none;
  padding: 10px;
  border-radius: 3px;
  margin-top: 10px;
  min-width: 125px;
  background-color: $color-red1;
  transition: 0.4s ease-in-out;
  cursor: pointer;
  text-align: left;
  color: $color-white1;
    .fa {
      margin-right: 10px;
      color: $color-white1;
    }

    &:hover {
      background-color: $color-red3;
    }
}

.nolike {
  background-color: $color-red3;
  &:hover {
    background-color: $color-red2;
  }
}

.popup {
  margin-top: 40px;
  display: flex;
  flex-direction: column;
  position: fixed;
  top: 25%;
  opacity: 0;
  background-color: $color-white1;
  transition: 0.1s ease-in-out;
  transform: translateY(-615px);
  box-shadow: 0 16px 40px 0 $color-black4;
  overflow: hidden;
  border-top: 2px solid $color-white1;
  z-index: -1;
  padding: 18px 55px;
  left: 16%;
  border-radius: 10px;
  overflow: auto;
  
}

.openpopup {
  width: 70%;
  opacity: 1;
  z-index: 20;
  transform: translateY(0);
}

.link-link-block {
  text-align: right;
  padding: 0px 0px;
  display: inline;
  width: 0%;
}

.linkinlike {
  background-image: url(../../assets/images/png/heart.png);
  background-repeat: no-repeat;
  cursor: pointer;
  padding: 15px 36px 2px 4px;
}
