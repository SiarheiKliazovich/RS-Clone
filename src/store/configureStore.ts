/* eslint-disable @typescript-eslint/ban-ts-comment */
import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";

import loggerMiddleware from "./logger";
import rootReducer from "../reducers";

const configureStore = () => {
  const middlewares = [loggerMiddleware, thunkMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, middlewareEnhancer);

  // @ts-ignore:next-line
  if (process.env.NODE_ENV !== "production" && module.hot) {
    // @ts-ignore:next-line
    module.hot.accept("./../reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
};

export default configureStore;
