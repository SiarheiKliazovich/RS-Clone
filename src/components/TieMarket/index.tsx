import { FunctionComponent, useState, useRef, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import classNames from "classnames";
import FormSignIn from "../FormSignIn";

import "./tiemarket.scss";
import { IStore } from "../../interfaces/store";
import { nightTheme } from "../../data/constants";
import useOnClickOutside from "../../hook/useOnClickOutside";
import Hover from "../Hover";
import tieMarketLang from "../../data/tieMarket";

import {
  tieFetching,
  tieFetched,
  tieFetchingError,
} from "../../actions/tieMarket/index";
import {
  buyTieFetching,
  buyTieFetched,
  buyTieFetchingError,
} from "../../actions/buyTie/index";
import { getTies } from "../../services/apiTies";
import { ITiesReducer, ITie } from "../../interfaces/tie";
import { ILangReducer } from "../../interfaces/langReducer";
import Spinner from "../Spinner";
import { IOrder } from "../../interfaces/order";
import { createOrder } from "../../services/apiOrders";

const TieMarket: FunctionComponent = () => {
  const backgroundColor = nightTheme.background.element;

  const saved = localStorage.getItem("favouriteTie");
  const [favouriteTies, setFavouriteTie] = useState<ITie[]>(
    saved ? JSON.parse(saved) : []
  );

  const [openPopup, setOpenPopup] = useState(false);
  const popupRef = useRef<HTMLDivElement>(null);
  useOnClickOutside(popupRef, () => setOpenPopup(!openPopup), openPopup);

  const { lang } = useSelector((state: ILangReducer) => state.langReducer);
  const tiesStore = useSelector((state: ITiesReducer) => state.tiesReducer);
  const interfaceSettings = useSelector((state: IStore) => state.appInterface);
  const { accentColor, isNavbarNightMode } = interfaceSettings;
  const { tieLoadingStatus, ties } = tiesStore;

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const listLang = tieMarketLang.find((c) => c.lang === lang)!;
  const dispatch = useDispatch();

  const buyTie = async (
    formData: Pick<IOrder, "userId" | "image" | "price" | "sellerId">
  ) => {
    try {
      dispatch(buyTieFetching());
      const login = localStorage.getItem("login");
      const user = login ? JSON.parse(login) : [];
      if (!user.user._id) {
        setOpenPopup(!openPopup);
      } else {
        const user = login ? JSON.parse(login) : [];
        const body = {
          ...formData,
          userId: user.user._id,
        };
        const resp = await createOrder(body);
        if (resp.userId) {
          dispatch(buyTieFetched(resp));
        }
      }
    } catch (e) {
      dispatch(buyTieFetchingError());
    }
  };

  const getTieList = async () => {
    try {
      dispatch(tieFetching());
      const ties = await getTies();
      dispatch(tieFetched(ties));
    } catch {
      dispatch(tieFetchingError());
    }
  };
  useEffect(() => {
    getTieList();
  }, []);

  const addFavouriteTie = (tie: ITie): void => {
    const favourite = favouriteTies.filter(
      (favouriteTie: ITie) => favouriteTie._id === tie._id
    );
    if (favourite.length > 0) {
      setFavouriteTie(
        favouriteTies.filter(
          (favouriteTie: ITie) => favouriteTie._id !== tie._id
        )
      );
    } else {
      setFavouriteTie([...favouriteTies, tie]);
    }
  };

  useEffect(() => {
    localStorage.setItem("favouriteTie", JSON.stringify(favouriteTies));
  }, [favouriteTies]);

  const spinner = tieLoadingStatus === "loading" ? <Spinner /> : null;

  return (
    <>
      <main className="main-market">
        <div className="tiemarket-wrapper">
          <div className="container">
            <div className="tiemarket__banner-wrapper">
              <div className="market-block">
                <div className="market-block__title_wrapper">
                  <h4 className="market-block__title">{listLang.data.title}</h4>
                  <div className="link-link-block">
                    <Link className="linkinlike" to={"/favourite-tie"}></Link>
                  </div>
                </div>
                <div className="market-block__products">
                  {spinner}
                  <div className="row__products">
                    {ties?.map((tie) => (
                      <div className="tie__product" key={tie.name}>
                        <div className="tie__products_img">
                          <img
                            className="products_img"
                            src={tie.image}
                            alt=""
                          />
                        </div>
                        <div className="tie__products_discription">
                          <p className="tie__products_name">{tie.name}</p>
                          <p className="tie__products_price">
                            {listLang.data.price} {tie.price}$
                          </p>
                          <div className="tie__products_wrapper">
                            <Hover>
                              <button
                                className="tie__products_btn"
                                onClick={() =>
                                  buyTie({
                                    ...tie,
                                    sellerId: tie.userId,
                                  })
                                }
                                style={{
                                  backgroundColor: isNavbarNightMode
                                    ? backgroundColor
                                    : accentColor.static,
                                }}
                              >
                                {listLang.data.btn}
                              </button>
                            </Hover>
                            <button
                              className={classNames("btn__like", {
                                nolike:
                                  favouriteTies.filter(
                                    (favouriteTie: ITie) =>
                                      favouriteTie._id === tie._id
                                  ).length !== 0,
                              })}
                              onClick={() => addFavouriteTie(tie)}
                            ></button>
                          </div>
                        </div>
                      </div>
                    ))}
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div className={classNames("popup-overlay", { overlay: openPopup })}>
          <div
            className={classNames("popup", { openpopup: openPopup })}
            ref={popupRef}
          >
            <FormSignIn />
          </div>
        </div>
      </main>
    </>
  );
};
export default TieMarket;
