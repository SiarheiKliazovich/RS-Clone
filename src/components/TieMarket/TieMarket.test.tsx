import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import TieMarket from ".";

const mockStore = configureStore([]);

describe("TieMarket", () => {
  // eslint-disable-next-line
  let store: any;

  beforeEach(() => {
    store = mockStore({
      langReducer: {
        lang: "en",
      },
      appInterface: {
        accentColor: {
          static: "#000",
          hover: "#fff",
        },
      },
      tiesReducer: {
        tieLoadingStatus: "idle",
      },
      buyTieReducer: {
        buyTieLoadingStatus: "load",
      },
      auth: {
        user: {
          role: "USER",
        },
      },
    });
  });

  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <TieMarket />
        </Router>
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
