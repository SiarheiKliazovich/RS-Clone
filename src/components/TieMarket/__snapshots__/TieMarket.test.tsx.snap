// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TieMarket renders correctly 1`] = `
<main
  class="main-market"
>
  <div
    class="tiemarket-wrapper"
    style="background-color: rgb(246, 246, 246);"
  >
    <div
      class="container"
    >
      <div
        class="tiemarket__banner-wrapper"
        style="background-color: rgb(246, 246, 246);"
      >
        <div
          class="market-block"
          style="background-color: rgb(255, 255, 255);"
        >
          <div
            class="market-block__title_wrapper"
          >
            <h4
              class="market-block__title"
            >
              Tie Market
            </h4>
          </div>
          <div
            class="market-block__products"
          >
            <div
              class="row__products"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="popup-overlay"
  >
    <div
      class="popup"
    >
      <div
        class="sign__content"
      >
        <h2
          class="sign__title"
        >
          sign.signIn
        </h2>
        <div
          class="sign__text"
        >
          sign.text-1
        </div>
        <form
          action="#"
          class="form"
        >
          <input
            class="field"
            id="email"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
          <input
            class="field"
            id="password"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <button
            class="sign__button"
            style="background-color: rgb(0, 0, 0);"
            type="submit"
          >
            sign.signIn
          </button>
        </form>
        <div
          class="sign__text-link"
        >
          sign.create
           
          <a
            class="text-link"
            href="/sign-up"
            style="color: rgb(0, 0, 0);"
          >
            sign.signUp
          </a>
        </div>
      </div>
    </div>
  </div>
</main>
`;
